tinkoff/invest/caching/market_data_cache/cache.py

```diff
@@ -0,0 +1,143 @@
+import logging
+from datetime import datetime, timedelta
+from typing import Dict, Generator, Iterable, Optional, Tuple
+
+from tinkoff.invest import CandleInterval, HistoricCandle
+from tinkoff.invest.caching.market_data_cache.cache_settings import (
+    MarketDataCacheSettings,
+)
+from tinkoff.invest.caching.market_data_cache.instrument_date_range_market_data import (
+    InstrumentDateRangeData,
+)
+from tinkoff.invest.caching.market_data_cache.instrument_market_data_storage import (
+    InstrumentMarketDataStorage,
+)
+from tinkoff.invest.candle_getter_interface import ICandleGetter
+from tinkoff.invest.services import Services
+from tinkoff.invest.utils import (
+    candle_interval_to_timedelta,
+    datetime_range_floor,
+    floor_datetime,
+    now,
+    with_filtering_distinct_candles,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class MarketDataCache(ICandleGetter):
+    def __init__(self, settings: MarketDataCacheSettings, services: Services):
+        self._settings = settings
+        self._settings.base_cache_dir.mkdir(parents=True, exist_ok=True)
+        self._services = services
+        self._figi_cache_storages: Dict[
+            Tuple[str, CandleInterval], InstrumentMarketDataStorage
+        ] = {}
+
+    def _get_candles_from_net(
+        self, figi: str, interval: CandleInterval, from_: datetime, to: datetime
+    ) -> Iterable[HistoricCandle]:
+        yield from self._services.get_all_candles(
+            figi=figi,
+            interval=interval,
+            from_=from_,
+            to=to,
+        )
+
+    def _with_saving_into_cache(
+        self,
+        storage: InstrumentMarketDataStorage,
+        from_net: Iterable[HistoricCandle],
+        net_range: Tuple[datetime, datetime],
+        interval_delta: timedelta,
+    ) -> Iterable[HistoricCandle]:
+        candles = list(from_net)
+        if candles:
+            filtered_net_range = self._round_net_range(net_range, interval_delta)
+            filtered_candles = list(self._filter_complete_candles(candles))
+            storage.update(
+                [
+                    InstrumentDateRangeData(
+                        date_range=filtered_net_range, historic_candles=filtered_candles
+                    )
+                ]
+            )
+            logger.debug("From net [\n%s\n%s\n]", str(net_range[0]), str(net_range[1]))
+            logger.debug(
+                "Filtered net [\n%s\n%s\n]",
+                str(filtered_net_range[0]),
+                str(filtered_net_range[1]),
+            )
+            logger.debug(
+                "Filtered net real [\n%s\n%s\n]",
+                str(min(list(map(lambda x: x.time, filtered_candles)))),  # noqa: C417
```

Comment:
C417: Unnecessary `map` usage - rewrite using a generator expression/`<list/set/dict>` comprehension.

`map(func, iterable)` has great performance when func is a built-in function, and it makes sense if your function already has a name. But if your func is a `lambda`, it’s faster to use a generator expression or a comprehension, as it avoids the function call overhead. For example:

- Rewrite `map(lambda x: x + 1, iterable)` to `(x + 1 for x in iterable)`
- Rewrite `map(lambda item: get_id(item), items)` to `(get_id(item) for item in items)`
- Rewrite `list(map(lambda num: num * 2, nums))` to `[num * 2 for num in nums]`
- Rewrite `set(map(lambda num: num % 2 == 0, nums))` to `{num % 2 == 0 for num in nums}`
- Rewrite `dict(map(lambda v: (v, v ** 2), values))` to `{v : v ** 2 for v in values}`